src/lib/hooks/swap/useSwapInfo.tsx

```diff
@@ -0,0 +1,157 @@
+import { Trans } from '@lingui/macro'
+import { Currency, CurrencyAmount, Percent, TradeType } from '@uniswap/sdk-core'
+import { useClientSideV3Trade } from 'hooks/useClientSideV3Trade'
+import { atom } from 'jotai'
+import { useAtomValue, useUpdateAtom } from 'jotai/utils'
+import { useCurrencyBalances } from 'lib/hooks/useCurrencyBalance'
+import { Field, swapAtom } from 'lib/state/swap'
+import tryParseCurrencyAmount from 'lib/utils/tryParseCurrencyAmount'
+import { ReactNode, useEffect, useMemo } from 'react'
+import { InterfaceTrade, TradeState } from 'state/routing/types'
+
+import { isAddress } from '../../../utils'
+import useActiveWeb3React from '../useActiveWeb3React'
+
+interface SwapInfo {
+  currencies: { [field in Field]?: Currency }
+  currencyBalances: { [field in Field]?: CurrencyAmount<Currency> }
+  currencyAmounts: { [field in Field]?: CurrencyAmount<Currency> }
+  trade: {
+    trade?: InterfaceTrade<Currency, Currency, TradeType>
+    state: TradeState
+  }
+  allowedSlippage: Percent
+}
+
+const BAD_RECIPIENT_ADDRESSES: { [address: string]: true } = {
+  '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f': true, // v2 factory
+  '0xf164fC0Ec4E93095b804a4795bBe1e041497b92a': true, // v2 router 01
+  '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D': true, // v2 router 02
+}
+
+// from the current swap inputs, compute the best trade and return it.
+function useComputeSwapInfo(): SwapInfo {
+  const { account } = useActiveWeb3React()
+
+  const {
+    independentField,
+    amount,
+    [Field.INPUT]: inputCurrency,
+    [Field.OUTPUT]: outputCurrency,
+  } = useAtomValue(swapAtom)
+
+  const to = account
+
+  const relevantTokenBalances = useCurrencyBalances(
+    account,
+    useMemo(() => [inputCurrency ?? undefined, outputCurrency ?? undefined], [inputCurrency, outputCurrency])
+  )
+
+  const isExactIn: boolean = independentField === Field.INPUT
+  const parsedAmount = useMemo(
+    () => tryParseCurrencyAmount(amount, (isExactIn ? inputCurrency : outputCurrency) ?? undefined),
+    [inputCurrency, isExactIn, outputCurrency, amount]
+  )
+
+  const trade = useClientSideV3Trade(
```

Comment:
nit: You could add a TODO on L55 that we'll need to eventually add a routing API hook here.